pkg/cloud/libvirt/actuators/machine/utils/cloudinit.go

```diff
@@ -53,28 +94,247 @@ func setCloudInit(domainDef *libvirtxml.Domain, cloudInit *providerconfigv1.Clou
-				Value: fmt.Sprintf("name=opt/actuator.libvirt.io.k8s.sigs/config%0*d,string=%s", len(chunks), idx, chunk),
-			})
-		}
+const userDataFileName string = "user-data"
+const metaDataFileName string = "meta-data"
+const networkConfigFileName string = "network-config"
+
+type defCloudInit struct {
+	Name          string
+	PoolName      string
+	MetaData      string `yaml:"meta_data"`
+	UserData      string `yaml:"user_data"`
+	NetworkConfig string `yaml:"network_config"`
+}
+
+func newCloudInitDef() defCloudInit {
+	return defCloudInit{}
+}
+
+// Create a ISO file based on the contents of the CloudInit instance and
+// uploads it to the libVirt pool
+// Returns a string holding terraform's internal ID of this resource
+func (ci *defCloudInit) CreateIso() (string, error) {
+	iso, err := ci.createISO()
+	if err != nil {
+		return "", err
+	}
+	return iso, err
+}
+
+// Create the ISO holding all the cloud-init data
+// Returns a string with the full path to the ISO file
+func (ci *defCloudInit) createISO() (string, error) {
+	log.Print("Creating new ISO")
+	tmpDir, err := ci.createFiles()
+	if err != nil {
+		return "", err
 	}
 
-	return nil
+	isoDestination := filepath.Join(tmpDir, ci.Name)
+	cmd := exec.Command(
+		"mkisofs",
+		"-output",
+		isoDestination,
+		"-volid",
+		"cidata",
+		"-joliet",
+		"-rock",
+		filepath.Join(tmpDir, userDataFileName),
+		filepath.Join(tmpDir, metaDataFileName),
+		filepath.Join(tmpDir, networkConfigFileName))
+
+	log.Printf("About to execute cmd: %+v", cmd)
+	if err = cmd.Run(); err != nil {
+		return "", fmt.Errorf("Error while starting the creation of CloudInit's ISO image: %s", err)
+	}
+	log.Printf("ISO created at %s", isoDestination)
+
+	return isoDestination, nil
 }
+
+// write user-data,  meta-data network-config in tmp files and dedicated directory
+// Returns a string containing the name of the temporary directory and an error
+// object
+func (ci *defCloudInit) createFiles() (string, error) {
+	log.Print("Creating ISO contents")
+	tmpDir, err := ioutil.TempDir("", "cloudinit")
+	if err != nil {
+		return "", fmt.Errorf("Cannot create tmp directory for cloudinit ISO generation: %s",
+			err)
+	}
+	// user-data
+	if err = ioutil.WriteFile(filepath.Join(tmpDir, userDataFileName), []byte(ci.UserData), os.ModePerm); err != nil {
+		return "", fmt.Errorf("Error while writing user-data to file: %s", err)
+	}
+	// meta-data
+	if err = ioutil.WriteFile(filepath.Join(tmpDir, metaDataFileName), []byte(ci.MetaData), os.ModePerm); err != nil {
+		return "", fmt.Errorf("Error while writing meta-data to file: %s", err)
+	}
+	// network-config
+	if err = ioutil.WriteFile(filepath.Join(tmpDir, networkConfigFileName), []byte(ci.NetworkConfig), os.ModePerm); err != nil {
+		return "", fmt.Errorf("Error while writing network-config to file: %s", err)
+	}
+
+	log.Print("ISO contents created")
+
+	return tmpDir, nil
+}
+
+func (ci *defCloudInit) UploadIso(client *Client, iso string) (string, error) {
```

Comment:
this is duplicating https://github.com/openshift/cluster-api-provider-libvirt/blob/master/pkg/cloud/libvirt/actuators/machine/utils/ignition.go#L48

Reply:
Not completely. There some differences like:
- `ign.createFile()` vs. `newImage(iso)`